hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationFuture.java

```diff
@@ -78,35 +79,42 @@ protected void onInterruptDetected() {
     protected E resolveAndThrowIfException(Object unresolved) throws ExecutionException, InterruptedException {
         Object value = resolve(unresolved);
 
-        if (value == null || !(value instanceof Throwable)) {
+        if (value == null || !(value instanceof AltResult)) {
             return (E) value;
-        } else if (value instanceof CancellationException) {
-            throw (CancellationException) value;
-        } else if (value instanceof ExecutionException) {
-            throw (ExecutionException) value;
-        } else if (value instanceof InterruptedException) {
-            throw (InterruptedException) value;
-        } else if (value instanceof Error) {
-            throw (Error) value;
+        }
+
+        Throwable cause = ((AltResult) value).cause;
+
+        if (cause instanceof CancellationException) {
+            throw (CancellationException) cause;
+        } else if (cause instanceof ExecutionException) {
+            throw (ExecutionException) cause;
+        } else if (cause instanceof InterruptedException) {
+            throw (InterruptedException) cause;
+        } else if (cause instanceof Error) {
+            throw (Error) cause;
         } else {
-            throw new ExecutionException((Throwable) value);
+            throw new ExecutionException(cause);
         }
     }
 
-    @SuppressWarnings("checkstyle:npathcomplexity")
+    @SuppressWarnings({"checkstyle:npathcomplexity", "checkstyle:cyclomaticcomplexity"})
     @Override
     protected Object resolve(Object unresolved) {
         if (unresolved == null) {
             return null;
         } else if (unresolved == INTERRUPTED) {
-            return new InterruptedException(invocation.op.getClass().getSimpleName() + " was interrupted. " + invocation);
+            return new AltResult(new InterruptedException(invocation.op.getClass().getSimpleName()
+                    + " was interrupted. " + invocation));
         } else if (unresolved == CALL_TIMEOUT) {
             return newOperationTimeoutException(false);
         } else if (unresolved == HEARTBEAT_TIMEOUT) {
             return newOperationTimeoutException(true);
         } else if (unresolved.getClass() == Packet.class) {
-            NormalResponse response = invocation.context.serializationService.toObject(unresolved);
-            unresolved = response.getValue();
+            unresolved = invocation.context.serializationService.toObject(unresolved);
+            if (unresolved instanceof NormalResponse) {
```

Comment:
In which case `unresolved ` is not an instance of `NormalResponse `?  What happens when this condition doesn't hold?

Reply:
When it is an instance of an ErrorResponse. So once it gets deserialized, it will proceed with processing it, outside the outer IF block.